json_rpc/src/request.rs

```diff
@@ -61,20 +55,76 @@ fn is_valid(id: &Value) -> Result<(), Error> {
     Ok(())
 }
 
-impl TryFrom<UnvalidatedRequest> for Request {
+impl TryFrom<Map<String, Value>> for Request {
     type Error = ErrorOrRejection;
 
     /// Returns `Ok` if the request is valid as per
     /// [the JSON-RPC specification](https://www.jsonrpc.org/specification#request_object).
     ///
-    /// Returns an `Error` in the following cases:
+    /// Returns an `Error` in any of the following cases:
     ///   * "jsonrpc" field is not "2.0"
+    ///   * "method" field is not a String
+    ///   * "params" field is present, but is not an Array or Object
     ///   * "id" field is not a String, valid Number or Null
     ///   * "id" field is a Number with fractional part
+    ///   * extra fields exist
     ///
     /// Returns a `Rejection` if the "id" field is `None`.
-    fn try_from(request: UnvalidatedRequest) -> Result<Self, Self::Error> {
-        let id = match request.id {
+    fn try_from(mut request: Map<String, Value>) -> Result<Self, Self::Error> {
+        // Just copy "id" field for now to return verbatim in any errors before we get to actually
+        // validating the "id" field itself.
+        let id = request.get(ID_FIELD_NAME).cloned().unwrap_or_default();
+
+        match request.remove(JSONRPC_FIELD_NAME) {
+            Some(Value::String(jsonrpc)) => {
+                if jsonrpc != JSON_RPC_VERSION {
+                    let error = Error::new(
+                        ReservedErrorCode::InvalidRequest,
+                        format!("'jsonrpc' must be '2.0', but was '{}'", jsonrpc),
```

Comment:
I'd phrase as
```suggestion
                        format!("expected 'jsonrpc' to be '2.0', got '{}'", jsonrpc),
```
Here and below. Not feeling strongly about this, up to you.

Reply:
Fixed in [a11b164](https://github.com/casper-network/casper-node/pull/3155/commits/a11b164bdc53b713db4f181e77bdcef32c33fae2).